packages/lib/SyncTargetJoplinServerSAML.ts

```
@@ -72,6 +72,35 @@ export default class SyncTargetJoplinServerSAML extends SyncTargetJoplinServer {
 		return false;
 	}
 
+	public static override async checkConfig(fileApi: FileApiOptions) {
+		try {
+			// Simulate a login request
+			const result = await fetch(`${fileApi.path()}/api/saml`);
+
+			if (result.status === 200) { // The server successfully responded, SAML is enabled
+				return {
+					ok: true,
+					errorMessage: '',
+				};
+			} else if (result.status === 403) { // The server responded with a forbidden response, SAML is disabled
+				return {
+					ok: false,
+					errorMessage: _('SAML is not enabled on this server.'),
+				};
+			} else { // Unknown error
+				return {
+					ok: false,
+					errorMessage: _('Failed to connect to the server'),
+				};
+			}
+		} catch (e) {
+			return {
+				ok: false,
+				errorMessage: e.message,
+			};
+		}
```

I feel the logic is more complex that it needs to be, and even hide the actual errors. Could I suggest something like this?

```
if (result.status === 200) { // The server successfully responded, SAML is enabled
	return {
		ok: true,
		errorMessage: '',
	};
} else {
	return {
		ok: false,
		errorMessage: `Could not connect to server: Error ${result.status}: ${result.statusText}`,
	};
}
```

Also error 403 could also happen in other conditions, isn't it? Like if the user's credentials are not valid? It seems strange that 403 is sent for something that's essentially down

---

A 403 error is returned by the server if SAML support is not enabled by the administrator, as per `packages/server/src/routes/api/login.ts`. This error code is not linked to the fact that the credentials are valid or not, since we never handle them when using SAML anyways (this is delegated to the identity provider). We only have access to the server URL at that point to perform the test, so I just check that the server seems to be reachable and that it is configured to accept a SAML authentication flow.

I tried your changes, but when the server is reachable but with SAML disabled, the client just shows "Could not connect to server: Error 403: Forbidden", which I think is not helpful when trying to diagnose an issue. I can instead show the whole response, but that will display raw JSON to the user ("Could not connect to server: Error 403: {"error":"SAML not enabled"}"), or I can parse that JSON and show only the `error` field, after checking that the response contains it. What do you think is the best here?

---

Thanks, I guess there are two separate issues:

- For SAML being disabled, I wonder if a better error would be `501 Not Implemented`? It's true that 403 is not returned for authentication in SAML context but it is in other contexts so I think it would make sense to use something else here.

- And indeed if there's a JSON response it should ideally be parsed and the "error" field should be displayed. We do this in other parts of the code.

---

I don't think that a 501 error is correct here. As per the [HTTP Semantics RFC on the 501 status code](https://www.rfc-editor.org/rfc/rfc9110#status.501):
> The 501 (Not Implemented) status code indicates that the server does not support the functionality required to fulfill the request. This is the appropriate response when the server does not recognize the request method and is not capable of supporting it for any resource.

In our case, the server does supports and recognizes the request, but the functionality was not enabled by the administrator. This is a case where I think that a 403 status code can be used, as documented by [the same RFC, on the section about the 403 code](https://www.rfc-editor.org/rfc/rfc9110#status.403):
> The 403 (Forbidden) status code indicates that the server understood the request but refuses to fulfill it. A server that wishes to make public why the request has been forbidden can describe that reason in the response content (if any).
>
> If authentication credentials were provided in the request, the server considers them insufficient to grant access. The client SHOULD NOT automatically repeat the request with the same credentials. The client MAY repeat the request with new or different credentials. However, a request might be forbidden for reasons unrelated to the credentials.

---

Not sure what you mean when you say the server does support and recognizes the request? It really doesn't if SAML is disabled, right? Just because there's a code path for the error doesn't mean it's supported. Or you mean it can potentially support it? But then it's not relevant in this case

---

When I say that it's supported and recognized, I meant that there is a code path for `/api/saml`. Even when SAML is disabled, we know that the request is about SAML, and the server refuses it because it was configured to do so. But I agree that this interpretation of the HTTP specification can be too pedantic, and we can deviate from that if you wish.

---

Ok, it's not hugely important so we can leave it as it is, that's definitely fine. On client side I still think there should only be two cases though: 200 and everything else + error message if available